### PR TITLE
feat: add support for dynamic matching

### PR DESCRIPTION
**Pull request by shink**
Fix: #169

### CODE REVIEW

**Comment by shy1st** [comment]
@shink Please add the test

**Comment by hsluoyz** [comment]
@fangzhengjin plz review.

**Comment by shink** [comment]
@hsluoyz @fangzhengjin @shy1st plz review.

**Review by mango-6bd — APPROVED** [review]

File: src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java [hunk 1]
```diff
@@ -344,14 +344,15 @@ public void buildRoleLinks() {
...
      *
-     * @param rvals the request needs to be mediated, usually an array
-     *              of strings, can be class instances if ABAC is used.
+     * @param matcher the custom matcher.
+     * @param rvals   the request needs to be mediated, usually an array
+     *                of strings, can be class instances if ABAC is used.
      * @return whether to allow the request.
      */
-    public boolean enforce(Object... rvals) {
+    private boolean enforce(String matcher, Object... rvals) {
```
Comment (hsluoyz) [inline src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java:355]: @shink why overload this function? Golang doesn't support overloading. And I want Java to follow the same function naming as Go, so even Java supports function overloading, we can choose not to use it actually.

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Review by fangzhengjin — APPROVED** [review]
LGTM

File: src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java (hunk 1 above)
Comment (shink) [reply src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java:355]: @hsluoyz There are two functions in Golang, [enforce](https://github.com/casbin/casbin/blob/c139c7702bd98bc8809a40127b3a9dbb937d3ea0/enforcer.go#L438) which is private and [Enforce](https://github.com/casbin/casbin/blob/c139c7702bd98bc8809a40127b3a9dbb937d3ea0/enforcer.go#L632) which is public. But the function name in java is usually lowercase, so I overload the `enforce` function.

Comment (hsluoyz) [reply src/main/java/org/casbin/jcasbin/main/CoreEnforcer.java:355]: OK

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**